misc/memory-connection-limits/tests/util.rs

```diff
@@ -0,0 +1,119 @@
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::task::{Context, Poll};
+
+use libp2p_core::{Endpoint, Multiaddr};
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    dummy, ConnectionClosed, ConnectionDenied, ConnectionId, FromSwarm, NetworkBehaviour,
+    PollParameters, THandler, THandlerInEvent, THandlerOutEvent, ToSwarm,
+};
+use void::Void;
+
+#[derive(libp2p_swarm_derive::NetworkBehaviour)]
+#[behaviour(prelude = "libp2p_swarm::derive_prelude")]
+pub(crate) struct TestBehaviour {
+    pub(crate) connection_limits: libp2p_memory_connection_limits::Behaviour,
+    pub(crate) mem: ConsumeMemoryBehaviour,
+}
+
+#[derive(Default)]
+pub(crate) struct ConsumeMemoryBehaviour {
+    map: Vec<Vec<u8>>,
+}
+
+impl ConsumeMemoryBehaviour {
+    fn handle_connection(&mut self) {
+        // 1MB
+        self.map.push(vec![1; 1024 * 1024]);
+    }
+}
+
+impl NetworkBehaviour for ConsumeMemoryBehaviour {
+    type ConnectionHandler = dummy::ConnectionHandler;
+    type ToSwarm = Void;
+
+    fn handle_pending_inbound_connection(
+        &mut self,
+        _: ConnectionId,
+        _: &Multiaddr,
+        _: &Multiaddr,
+    ) -> Result<(), ConnectionDenied> {
+        self.handle_connection();
+        Ok(())
+    }
+
+    fn handle_pending_outbound_connection(
+        &mut self,
+        _: ConnectionId,
+        _: Option<PeerId>,
+        _: &[Multiaddr],
+        _: Endpoint,
+    ) -> Result<Vec<Multiaddr>, ConnectionDenied> {
+        self.handle_connection();
+        Ok(vec![])
+    }
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        _: ConnectionId,
+        _: PeerId,
+        _: &Multiaddr,
+        _: &Multiaddr,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        self.handle_connection();
+        Ok(dummy::ConnectionHandler)
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        _: ConnectionId,
+        _: PeerId,
+        _: &Multiaddr,
+        _: Endpoint,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        self.handle_connection();
+        Ok(dummy::ConnectionHandler)
+    }
```

Comment:
By doing this in all 4 handlers, you are allocating 2 MB per connection, I think doing it only for each established is good enough :)

Reply:
In these tests, connections are never established, I've also changed mem consumption for established to 0